File: libraries/python/src/urc_converter/sources/bms/parse.py
"""Parser for BMS-family bytes."""

from ...error import UrcError
from .model import BmsChart


class _JavaRandom:
	"""java.util.Random-compatible 48-bit LCG (for the #RANDOM seed option)."""

	_MASK = (1 << 48) - 1
	_MULT = 0x5DEECE66D
	_ADD = 0xB

	def __init__(self, seed: int) -> None:
		self._state = (seed ^ self._MULT) & self._MASK

	def _next(self, bits: int) -> int:
		self._state = (self._state * self._MULT + self._ADD) & self._MASK
		return self._state >> (48 - bits)

	def next_int(self, bound: int) -> int:
		if bound & (-bound) == bound:
			return (bound * self._next(31)) >> 31
		while True:
			bits = self._next(31)
			value = bits % bound
			if bits - value + (bound - 1) <= 0x7FFFFFFF:
				return value


def parse_bms(
	data: bytes,
	*,
	pms: bool = False,
	seed: int | None = None,
	branches: list[int] | None = None,
) -> BmsChart:
	"""Parse BMS-family bytes into a source model."""
	text = _decode(data)
	return _parse(text, _scan_base(text), pms, seed, branches)


def _parse(
	text: str,
	base: int,
	pms: bool,
	seed: int | None,
	branches: list[int] | None,
) -> BmsChart:
	chart = BmsChart(pms=pms, base=base)
	random_gen = _JavaRandom(seed) if seed is not None else None
	frames: list[list[bool]] = []
	random_value: int | None = None
	branch_index = 0

	for line_no, raw in enumerate(_lines(text), start=1):
		line = raw.strip()
		if not line.startswith("#"):
			continue
		head = line[1:]

		if len(head) >= 6 and head[:3].isdigit() and head[5] == ":":
			if not any(frame[0] for frame in frames):
				_add_message(chart, int(head[:3]), head[3:5], head[6:], line_no)
			continue

		command, _, argument = head.partition(" ")
		word = command.upper()

		if word == "RANDOM":
			count = _to_int(argument, line_no, "#RANDOM")
			if count < 1:
				raise UrcError("syntax", line_no, f"#RANDOM count must be >= 1: {count}")
			if branches is not None and branch_index < len(branches):
				pick = branches[branch_index]
				if not 1 <= pick <= count:
					raise UrcError("syntax", line_no, f"branch pick out of range: {pick}")
			elif random_gen is not None:
				pick = random_gen.next_int(count) + 1
			else:
				pick = 1
			random_value = pick
			branch_index += 1
		elif word == "IF":
			if random_value is None:
				raise UrcError("syntax", line_no, "unmatched #IF")
			condition = _to_int(argument, line_no, "#IF")
			frames.append([random_value != condition, random_value == condition])
		elif word == "ELSEIF":
			if not frames:
				raise UrcError("syntax", line_no, "unmatched #ELSEIF")
			condition = _to_int(argument, line_no, "#ELSEIF")
			matched = frames[-1][1] or random_value == condition
			frames[-1] = [not matched, matched]
		elif word == "ELSE":
			if not frames:
				raise UrcError("syntax", line_no, "unmatched #ELSE")
			frames[-1] = [frames[-1][1], True]
		elif word == "ENDIF":
			if not frames:
				raise UrcError("syntax", line_no, "unmatched #ENDIF")
			frames.pop()
		elif word in (
			"SETRANDOM",
			"ENDRANDOM",
			"SWITCH",
			"CASE",
			"SKIP",
			"DEF",
			"ENDSW",
			"SETSWITCH",
		):
			raise UrcError("unsupported-version", line_no, f"unsupported BMS command: #{command}")
		elif any(frame[0] for frame in frames):
			continue
		elif word == "BPM":
			chart.bpm = _to_float(argument, line_no, "#BPM")
		elif (len(command) == 5 and command[:3].upper() == "BPM") or (
			len(command) == 8 and command[:6].upper() == "EXBPM"
		):
			chart.bpm_defs[command[-2:]] = _to_float(argument, line_no, command)
		elif len(command) == 6 and command[:4].upper() == "STOP":
			chart.stop_defs[command[-2:]] = abs(_to_float(argument, line_no, command)) / 192.0
		elif len(command) == 8 and command[:6].upper() == "SCROLL":
			chart.scroll_defs[command[-2:]] = _to_float(argument, line_no, command)
		elif word == "LNTYPE":
			chart.lntype = _to_int(argument, line_no, "#LNTYPE")
			if chart.lntype not in (1, 2):
				raise UrcError("syntax", line_no, f"unsupported #LNTYPE: {chart.lntype}")
		elif word == "LNOBJ":
			chart.lnobj = argument.strip() or None
		elif word == "TITLE":
			chart.title = argument.strip() or None
		elif word == "ARTIST":
			chart.artist = argument.strip() or None
		elif word == "PLAYLEVEL":
			chart.play_level = argument.strip() or None

	if frames:
		raise UrcError("syntax", 1, "unterminated #IF block")
	return chart


def _add_message(
	chart: BmsChart, measure: int, channel: str, payload: str, line_no: int
) -> None:
	if channel == "02":
		rate = _to_float(payload, line_no, "measure length")
		if rate < 0:
			raise UrcError("syntax", line_no, f"negative measure length: {payload}")
		chart.rates[measure] = rate
		return
	if len(payload) % 2 != 0 or not all(_is_id_char(char) for char in payload):
		raise UrcError("syntax", line_no, f"malformed object list: {payload!r}")
	ids = [payload[index : index + 2] for index in range(0, len(payload), 2)]
	chart.measures.setdefault(measure, {}).setdefault(channel, []).extend(ids)


def _decode(data: bytes) -> str:
	if data.startswith(b"\xef\xbb\xbf"):
		data = data[3:]
	for encoding in ("utf-8", "cp932"):
		try:
			return data.decode(encoding)
		except UnicodeDecodeError:
			continue
	raise UrcError("syntax", 1, "undecodable bytes: expected UTF-8 or Shift_JIS")


def _scan_base(text: str) -> int:
	for raw in _lines(text):
		parts = raw.strip().split()
		if len(parts) == 2 and parts[0].upper() == "#BASE":
			try:
				base = int(parts[1])
			except ValueError:
				raise UrcError("syntax", 1, f"invalid #BASE: {parts[1]}") from None
			if base not in (36, 62):
				raise UrcError("syntax", 1, f"unsupported #BASE: {base}")
			return base
	return 36


def _lines(text: str) -> list[str]:
	if text.startswith("﻿"):
		text = text[1:]
	return text.replace("\r\n", "\n").replace("\r", "\n").split("\n")


def _is_id_char(char: str) -> bool:
	return ("0" <= char <= "9") or ("A" <= char <= "Z") or ("a" <= char <= "z")


def _to_int(value: str, line_no: int, name: str) -> int:
	try:
		return int(value.strip())
	except ValueError:
		raise UrcError("syntax", line_no, f"invalid {name}: {value!r}") from None


def _to_float(value: str, line_no: int, name: str) -> float:
	try:
		return float(value.strip())
	except ValueError:
		raise UrcError("syntax", line_no, f"invalid {name}: {value!r}") from None

File: libraries/dotnet/UrcConverter/Sources/Bms/Parse.fs
namespace UrcConverter.Sources.Bms

module Parse =

    open System
    open System.Globalization
    open System.Text
    open FsToolkit.ErrorHandling
    open UrcConverter
    open UrcConverter.Sources.Bms.Model

    type ParseOptions =
        {
            Pms: bool
            Seed: int64 option
            Branches: uint64 list option
        }

    type private JavaRandom =
        {
            mutable State: uint64
        }

    module private JavaRandom =
        let mask = (1UL <<< 48) - 1UL
        let mult = 0x5DEECE66DUL
        let add = 0xBUL

        let create (seed: int64) : JavaRandom =
            { State = ((uint64 seed) ^^^ mult) &&& mask }

        let next (rng: JavaRandom) (bits: int) : uint64 =
            rng.State <- ((rng.State * mult) + add) &&& mask
            rng.State >>> (48 - bits)

        let nextInt (rng: JavaRandom) (bound: uint64) : uint64 =
            if (bound &&& (~~~bound + 1UL)) = bound then
                (bound * (next rng 31)) >>> 31
            else
                let rec loop () =
                    let bits = next rng 31
                    let value = bits % bound
                    if bits - value + (bound - 1UL) <= 0x7FFFFFFFUL then
                        value
                    else
                        loop ()
                loop ()

    let private decode (bytes: byte[]) : Result<string, UrcError> =
        Encoding.RegisterProvider(CodePagesEncodingProvider.Instance)
        let cleanBytes =
            if bytes.Length >= 3 && bytes[0] = 0xefuy && bytes[1] = 0xbbuy && bytes[2] = 0xbfuy then
                bytes[3..]
            else
                bytes

        let utf8 = UTF8Encoding(false, true)
        try
            Result.Ok(utf8.GetString(cleanBytes))
        with _ ->
            try
                let sjis = Encoding.GetEncoding("shift-jis", EncoderFallback.ExceptionFallback, DecoderFallback.ExceptionFallback)
                Result.Ok(sjis.GetString(cleanBytes))
            with _ ->
                Result.Error(UrcError.Syntax(1, "undecodable bytes: expected UTF-8 or Shift_JIS"))

    let private scanBase (text: string) : Result<int, UrcError> =
        let lines = text.Split([| "\r\n"; "\n"; "\r" |], StringSplitOptions.None)
        let rec loop (idx: int) =
            if idx >= lines.Length then
                Result.Ok 36
            else
                let parts = lines[idx].Trim().Split([| ' '; '\t' |], StringSplitOptions.RemoveEmptyEntries)
                if parts.Length = 2 && String.Equals(parts[0], "#BASE", StringComparison.OrdinalIgnoreCase) then
                    match Int32.TryParse(parts[1]) with
                    | true, b when b = 36 || b = 62 -> Result.Ok b
                    | true, b -> Result.Error(UrcError.Syntax(1, $"unsupported #BASE: {b}"))
                    | false, _ -> Result.Error(UrcError.Syntax(1, $"invalid #BASE: {parts[1]}"))
                else
                    loop (idx + 1)
        loop 0

    let private isIdChar (c: char) =
        (c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z')

    let private toInt (value: string) (lineNo: int) (name: string) : Result<int, UrcError> =
        match Int32.TryParse(value.Trim()) with
        | true, v -> Result.Ok v
        | false, _ -> Result.Error(UrcError.Syntax(lineNo, $"invalid {name}: {value}"))

    let private toFloat (value: string) (lineNo: int) (name: string) : Result<float, UrcError> =
        match Double.TryParse(value.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture) with
        | true, v -> Result.Ok v
        | false, _ -> Result.Error(UrcError.Syntax(lineNo, $"invalid {name}: {value}"))

    let private addMessage
        (chart: BmsChart)
        (measure: int)
        (channel: string)
        (payload: string)
        (lineNo: int)
        : Result<BmsChart, UrcError> =
        result {
            if channel = "02" then
                let! rate = toFloat payload lineNo "measure length"
                if rate < 0.0 then
                    return! Result.Error(UrcError.Syntax(lineNo, $"negative measure length: {payload}"))
                return { chart with Rates = Map.add measure rate chart.Rates }
            else
                if payload.Length % 2 <> 0 || not (payload |> Seq.forall isIdChar) then
                    return! Result.Error(UrcError.Syntax(lineNo, $"malformed object list: \"{payload}\""))
                let ids = [ for i in 0 .. 2 .. payload.Length - 2 -> payload.Substring(i, 2) ]
                let measureMap = Map.tryFind measure chart.Measures |> Option.defaultValue Map.empty
                let channelList = Map.tryFind channel measureMap |> Option.defaultValue []
                let updatedChannelList = channelList @ ids
                let updatedMeasureMap = Map.add channel updatedChannelList measureMap
                return { chart with Measures = Map.add measure updatedMeasureMap chart.Measures }
        }

    let parseBms (data: byte[]) (options: ParseOptions) : Result<BmsChart, UrcError> =
        result {
            let! text = decode data
            let! b = scanBase text
            let lines = text.TrimStart('\uFEFF').Split([| "\r\n"; "\n"; "\r" |], StringSplitOptions.None)

            let mutable chart =
                {
                    Pms = options.Pms
                    Base = b
                    Title = None
                    Artist = None
                    PlayLevel = None
                    Bpm = None
                    LnType = 1
                    LnObj = None
                    BpmDefs = Map.empty
                    StopDefs = Map.empty
                    ScrollDefs = Map.empty
                    Rates = Map.empty
                    Measures = Map.empty
                }

            let rng = options.Seed |> Option.map JavaRandom.create
            let mutable frames: (bool * bool) list = []
            let mutable randomValue: uint64 option = None
            let mutable branchIndex = 0

            for i in 0 .. lines.Length - 1 do
                let lineNo = i + 1
                let line = lines[i].Trim()
                if line.StartsWith("#") then
                    let head = line.Substring(1)
                    if head.Length >= 6 && Char.IsDigit(head[0]) && Char.IsDigit(head[1]) && Char.IsDigit(head[2]) && head[5] = ':' then
                        if not (frames |> List.exists fst) then
                            let measure = Int32.Parse(head.Substring(0, 3))
                            let channel = head.Substring(3, 2)
                            let payload = head.Substring(6)
                            let! updated = addMessage chart measure channel payload lineNo
                            chart <- updated
                    else
                        let spaceIdx = head.IndexOf(' ')
                        let command = if spaceIdx = -1 then head else head.Substring(0, spaceIdx)
                        let argument = if spaceIdx = -1 then "" else head.Substring(spaceIdx + 1).Trim()
                        let word = command.ToUpperInvariant()

                        match word with
                        | "RANDOM" ->
                            let! count = toInt argument lineNo "#RANDOM"
                            if count < 1 then
                                return! Result.Error(UrcError.Syntax(lineNo, $"#RANDOM count must be >= 1: {count}"))
                            let! pick =
                                match options.Branches with
                                | Some branches when branchIndex < branches.Length ->
                                    let p = branches[branchIndex]
                                    if p < 1UL || p > uint64 count then
                                        Result.Error(UrcError.Syntax(lineNo, $"branch pick out of range: {p}"))
                                    else
                                        Result.Ok p
                                | _ ->
                                    match rng with
                                    | Some r -> Result.Ok (JavaRandom.nextInt r (uint64 count) + 1UL)
                                    | None -> Result.Ok 1UL
                            randomValue <- Some pick
                            branchIndex <- branchIndex + 1
                        | "IF" ->
                            let! v =
                                randomValue
                                |> Option.map Result.Ok
                                |> Option.defaultValue (Result.Error(UrcError.Syntax(lineNo, "unmatched #IF")))
                            let! cond = toInt argument lineNo "#IF"
                            frames <- (v <> uint64 cond, v = uint64 cond) :: frames
                        | "ELSEIF" ->
                            match frames with
                            | [] -> return! Result.Error(UrcError.Syntax(lineNo, "unmatched #ELSEIF"))
                            | (active, matched) :: rest ->
                                let! cond = toInt argument lineNo "#ELSEIF"
                                let v = randomValue.Value
                                let nowMatched = matched || v = uint64 cond
                                frames <- (not nowMatched, nowMatched) :: rest
                        | "ELSE" ->
                            match frames with
                            | [] -> return! Result.Error(UrcError.Syntax(lineNo, "unmatched #ELSE"))
                            | (active, matched) :: rest ->
                                frames <- (matched, true) :: rest
                        | "ENDIF" ->
                            match frames with
                            | [] -> return! Result.Error(UrcError.Syntax(lineNo, "unmatched #ENDIF"))
                            | _ :: rest -> frames <- rest
                        | "SETRANDOM" | "ENDRANDOM" | "SWITCH" | "CASE" | "SKIP" | "DEF" | "ENDSW" | "SETSWITCH" ->
                            return! Result.Error(UrcError.UnsupportedVersion(lineNo, $"unsupported BMS command: #{command}"))
                        | _ when frames |> List.exists fst -> ()
                        | "BPM" ->
                            let! bpm = toFloat argument lineNo "#BPM"
                            chart <- { chart with Bpm = Some bpm }
                        | _ when (command.Length = 5 && command.StartsWith("BPM", StringComparison.OrdinalIgnoreCase))
                              || (command.Length = 8 && command.StartsWith("EXBPM", StringComparison.OrdinalIgnoreCase)) ->
                            let! bpm = toFloat argument lineNo command
                            chart <- { chart with BpmDefs = Map.add (command.Substring(command.Length - 2)) bpm chart.BpmDefs }
                        | _ when command.Length = 6 && command.StartsWith("STOP", StringComparison.OrdinalIgnoreCase) ->
                            let! stop = toFloat argument lineNo command
                            chart <- { chart with StopDefs = Map.add (command.Substring(command.Length - 2)) (abs stop / 192.0) chart.StopDefs }
                        | _ when command.Length = 8 && command.StartsWith("SCROLL", StringComparison.OrdinalIgnoreCase) ->
                            let! scroll = toFloat argument lineNo command
                            chart <- { chart with ScrollDefs = Map.add (command.Substring(command.Length - 2)) scroll chart.ScrollDefs }
                        | "LNTYPE" ->
                            let! lntype = toInt argument lineNo "#LNTYPE"
                            if lntype <> 1 && lntype <> 2 then
                                return! Result.Error(UrcError.Syntax(lineNo, $"unsupported #LNTYPE: {lntype}"))
                            chart <- { chart with LnType = lntype }
                        | "LNOBJ" ->
                            chart <- { chart with LnObj = if String.IsNullOrEmpty argument then None else Some argument }
                        | "TITLE" ->
                            chart <- { chart with Title = if String.IsNullOrEmpty argument then None else Some argument }
                        | "ARTIST" ->
                            chart <- { chart with Artist = if String.IsNullOrEmpty argument then None else Some argument }
                        | "PLAYLEVEL" ->
                            chart <- { chart with PlayLevel = if String.IsNullOrEmpty argument then None else Some argument }
                        | _ -> ()

            if not (List.isEmpty frames) then
                return! Result.Error(UrcError.Syntax(1, "unterminated #IF block"))
            return chart
        }
